// huawei-wingle-4g/lib/src/page/sms/SmsInDraft.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const log4js_1 = __importDefault(require("log4js"));
const StringUtils_1 = require("../../utils/StringUtils");
const Connection_1 = __importDefault(require("../../connection/Connection"));
class default_1 {
    constructor(login) {
        this.login = login;
        this.connection = login.getConnnection();
        this.logger = log4js_1.default.getLogger(StringUtils_1.substringAfter(__filename, 'huawei-wingle-4g'));
    }
    activeLog(activeLog) {
        this.login.activeLog(activeLog);
        this.logger.level = activeLog ? 'debug' : 'OFF';
    }
    saveSmsInDraft(phoneNumbers, content) {
        return __awaiter(this, void 0, void 0, function* () {
            const parameters = this.buildParameters(phoneNumbers, content);
            yield this.connection.openHomePage();
            yield this.login.login();
            const response = yield this.connection.post('/api/sms/save-sms', parameters);
            if (Connection_1.default.isSuccess(response)) {
                this.logger.debug('Saving SMS in draft done with success');
            }
            else {
                throw new Error('Unabel to save SMS in draft');
            }
        });
    }
    buildParameters(phoneNumbers, content) {
        let numbers = [];
        if (Array.isArray(phoneNumbers)) {
            phoneNumbers.forEach(phoneNumber => numbers.push(phoneNumber));
        }
        else {
            numbers.push(phoneNumbers);
        }
        numbers = numbers.filter(phone => phone);
        if (!numbers.length) {
            throw new Error('No phone number provided');
        }
        if (!content) {
            throw new Error('No content SMS provided');
        }
        if (content.length > 160) {
            throw new Error('Content SMS length cannot be upper than 160');
        }
        const phones = numbers.map(number => `<Phone>${number}</Phone>`).join('');
        const date = this.getDate();
        return `<?xml version: "1.0" encoding="UTF-8"?><request><Index>-1</Index><Phones>${phones}</Phones><Sca></Sca><Content>${content}</Content><Length>${content.length}</Length><Reserved>1</Reserved><Date>${date}</Date></request>`;
    }
    getDate() {
        const date = new Date();
        const year = date.getFullYear();
        const month = StringUtils_1.toTwo(date.getMonth() + 1);
        const day = StringUtils_1.toTwo(date.getDate());
        const hours = StringUtils_1.toTwo(date.getHours());
        const minutes = StringUtils_1.toTwo(date.getMinutes());
        const secondes = StringUtils_1.toTwo(date.getSeconds());
        return `${year}-${month}-${day} ${hours}:${minutes}:${secondes}`;
    }
}
exports.default = default_1;
