// huawei-wingle-4g/lib/src/page/sms/SmsRemover.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Connection_1 = __importDefault(require("../../connection/Connection"));
const log4js_1 = __importDefault(require("log4js"));
const StringUtils_1 = require("../../utils/StringUtils");
class default_1 {
    constructor(login) {
        this.login = login;
        this.connection = login.getConnnection();
        this.logger = log4js_1.default.getLogger(StringUtils_1.substringAfter(__filename, 'huawei-wingle-4g'));
    }
    activeLog(activeLog) {
        this.login.activeLog(activeLog);
        this.logger.level = activeLog ? 'debug' : 'OFF';
    }
    removeSms(smsIds) {
        return __awaiter(this, void 0, void 0, function* () {
            const parameters = this.buildParameters(smsIds);
            yield this.connection.openHomePage();
            yield this.login.login();
            const response = yield this.connection.post('/api/sms/delete-sms', parameters);
            if (Connection_1.default.isSuccess(response)) {
                this.logger.debug('Removing SMS done with success');
            }
            else {
                throw new Error(`Unable to remove SMS`);
            }
        });
    }
    buildParameters(smsIds) {
        let ids = [];
        if (Array.isArray(smsIds)) {
            smsIds.forEach(smsId => ids.push(smsId));
        }
        else {
            ids.push(smsIds);
        }
        ids = ids.filter(id => id);
        if (!ids.length) {
            throw new Error('No SMS id provided');
        }
        const indices = ids.map(id => `<Index>${id}</Index>`).join('');
        return `<?xml version: "1.0" encoding="UTF-8"?><request>${indices}</request>`;
    }
}
exports.default = default_1;
