// huawei-wingle-4g/lib/src/page/statistics/HistoryCleaner.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Connection_1 = __importDefault(require("../../connection/Connection"));
const log4js_1 = __importDefault(require("log4js"));
const StringUtils_1 = require("../../utils/StringUtils");
class default_1 {
    constructor(login) {
        this.login = login;
        this.connection = login.getConnnection();
        this.logger = log4js_1.default.getLogger(StringUtils_1.substringAfter(__filename, 'huawei-wingle-4g'));
    }
    activeLog(activeLog) {
        this.logger.level = activeLog ? 'debug' : 'OFF';
        this.login.activeLog(activeLog);
    }
    clearHistory() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.connection.openHomePage();
            yield this.login.login();
            const parameters = '<?xml version: "1.0" encoding="UTF-8"?><request><ClearTraffic>1</ClearTraffic></request>';
            const response = yield this.connection.post('/api/monitoring/clear-traffic', parameters);
            if (Connection_1.default.isSuccess(response)) {
                this.logger.debug('Stastistics history clear with success');
            }
            else {
                throw new Error('Unable to clear statistics history');
            }
        });
    }
}
exports.default = default_1;
