// huawei-wingle-4g/lib/src/page/statistics/WlanClientsExtractor.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const log4js_1 = __importDefault(require("log4js"));
const StringUtils_1 = require("../../utils/StringUtils");
const ElemenUtils_1 = __importDefault(require("../../utils/ElemenUtils"));
class default_1 {
    constructor(login) {
        this.login = login;
        this.connection = login.getConnnection();
        this.logger = log4js_1.default.getLogger(StringUtils_1.substringAfter(__filename, 'huawei-wingle-4g'));
    }
    activeLog(activeLog) {
        this.logger.level = activeLog ? 'debug' : 'OFF';
        this.login.activeLog(activeLog);
    }
    getWlanClients() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.connection.openHomePage();
            yield this.login.login();
            const response = yield this.connection.get('/api/wlan/host-list');
            const hostElements = response.document.querySelectorAll('Hosts > Host');
            const wlanClients = [];
            if (hostElements && hostElements.length) {
                hostElements.forEach(hostElement => {
                    const id = ElemenUtils_1.default(hostElement, 'ID', 'ID client', this.logger);
                    const ipAddress = ElemenUtils_1.default(hostElement, 'IpAddress', 'IP address', this.logger);
                    const hostName = ElemenUtils_1.default(hostElement, 'HostName', 'Host name', this.logger);
                    const macAddress = ElemenUtils_1.default(hostElement, 'MacAddress', 'Mac address', this.logger);
                    const duration = this.getDuration(hostElement);
                    wlanClients.push({ id, ipAddress, hostName, macAddress, duration });
                });
            }
            return wlanClients;
        });
    }
    getDuration(hostElement) {
        const rawDuration = ElemenUtils_1.default(hostElement, 'AssociatedTime', 'Duration', this.logger);
        return +rawDuration * 1000;
    }
}
exports.default = default_1;
