// huawei-wingle-4g/lib/src/page/home/Home.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Home = void 0;
const NetworkExtractor_1 = __importDefault(require("./NetworkExtractor"));
const CurrentConnectionExtractor_1 = __importDefault(require("./CurrentConnectionExtractor"));
const WlanStatusExtractor_1 = __importDefault(require("./WlanStatusExtractor"));
const DataMobile_1 = __importDefault(require("./DataMobile"));
class Home {
    constructor(login) {
        this.networkExtractor = new NetworkExtractor_1.default(login.getConnnection());
        this.currentConnectionExtractor = new CurrentConnectionExtractor_1.default(login.getConnnection());
        this.wlanStatusExtractor = new WlanStatusExtractor_1.default(login.getConnnection());
        this.dataMobile = new DataMobile_1.default(login);
    }
    activeLog(activeLog) {
        this.networkExtractor.activeLog(activeLog);
        this.currentConnectionExtractor.activeLog(activeLog);
        this.wlanStatusExtractor.activeLog(activeLog);
        this.dataMobile.activeLog(activeLog);
    }
    getNetwork() {
        return this.networkExtractor.getNetwork();
    }
    getCurrentConnection() {
        return this.currentConnectionExtractor.getCurrentConnection();
    }
    getWlanInformation() {
        return this.wlanStatusExtractor.getWlanInformation();
    }
    connectDataMobile() {
        return this.dataMobile.connect();
    }
    disconnectDataMobile() {
        return this.dataMobile.disconnect();
    }
}
exports.Home = Home;
